src/AccessToken.php

```diff
@@ -0,0 +1,308 @@
+
+        $this->httpHandler = $httpHandler
+            ?: HttpHandlerFactory::build(HttpClientCache::getHttpClient());
+        $this->cache = $cache ?: new MemoryCacheItemPool();
+        $this->configureJwtService();
+
+        // set phpseclib constants if applicable
+        $this->setPhpsecConstants();
+    }
+
+    /**
+     * Verifies an id token and returns the authenticated apiLoginTicket.
+     * Throws an exception if the id token is not valid.
+     * The audience parameter can be used to control which id tokens are
+     * accepted.  By default, the id token must have been issued to this OAuth2 client.
+     *
+     * @param string $token The JSON Web Token to be verified.
+     * @param array $options [optional] {
+     *     Configuration options.
+     *
+     *     @type string $audience The indended recipient of the token.
+     *     @type string $certsLocation The location (remote or local) from which
+     *        to retrieve certificates, if not cached. This value should only be
+     *        provided in limited circumstances in which you are sure of the
+     *        behavior.
+     * }
+     * @return array|bool the token payload, if successful, or false if not.
+     */
+    public function verify($token, array $options = [])
+    {
+        $options += [
+            'audience' => null,
+            'certsLocation' => self::FEDERATED_SIGNON_CERT_URL,
+        ];
+
+        // Check signature against each available cert.
+        // allow the loop to complete unless a known bad result is encountered.
+        $certs = $this->getFederatedSignOnCerts($options['certsLocation'], $options);
+        foreach ($certs as $cert) {
+            $rsa = new RSA();
+            $rsa->loadKey([
+                'n' => new BigInteger($this->callJwtStatic('urlsafeB64Decode', [
+                    $cert['n']
+                ]), 256),
+                'e' => new BigInteger($this->callJwtStatic('urlsafeB64Decode', [
+                    $cert['e']
+                ]), 256)
+            ]);
+
+            try {
+                $pubkey = $rsa->getPublicKey();
+                $payload = $this->callJwtStatic('decode', [
+                    $token,
+                    $pubkey,
+                    ['RS256']
+                ]);
+
+                if (property_exists($payload, 'aud')) {
+                    if ($options['audience'] && $payload->aud != $options['audience']) {
+                        return false;
+                    }
+                }
+
+                // support HTTP and HTTPS issuers
+                // @see https://developers.google.com/identity/sign-in/web/backend-auth
+                $issuers = array(self::OAUTH2_ISSUER, self::OAUTH2_ISSUER_HTTPS);
+                if (!isset($payload->iss) || !in_array($payload->iss, $issuers)) {
+                    return false;
+                }
+
+                return (array) $payload;
+            } catch (ExpiredException $e) {
+                return false;
```

Comment:
No, because `ExpiredException` means that the decryption succeeded but the token was expired. The other catches indicate that the token could not be decrypted *with the current key*. There may be other keys, so catching and doing nothing allows the loop over `$certs` to attempt decryption with a different key, if more are available. Only after each key has been attempted does the method return false.

Reply:
Makes sense, I looked too deep and the outside context of the  loop 👍